state-chain/runtime/src/migrations/housekeeping/clear_solana_egress_witnessing_elections.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2025 Chainflip Labs GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::*;
+use cf_runtime_utilities::genesis_hashes;
+#[cfg(feature = "try-runtime")]
+use codec::{Decode, Encode};
+use frame_support::{pallet_prelude::Weight, traits::OnRuntimeUpgrade};
+
+use pallet_cf_elections::{
+	electoral_system_runner::RunnerStorageAccessTrait,
+	electoral_systems::composite::tuple_7_impls::CompositeElectionIdentifierExtra,
+	ElectionProperties, RunnerStorageAccess,
+};
+#[cfg(feature = "try-runtime")]
+use sp_runtime::DispatchError;
+
+pub struct ClearSolanaEgressWitnessingElections;
+
+impl OnRuntimeUpgrade for ClearSolanaEgressWitnessingElections {
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, sp_runtime::TryRuntimeError> {
+		let count_egress_witnessing: u64 = ElectionProperties::<Runtime, SolanaInstance>::iter_keys()
+			.filter(|id| matches!(id.extra(), CompositeElectionIdentifierExtra::D(_)))
+			.count()
+			.try_into()
+			.unwrap();
+		log::info!("🧹 Found {} EgressWitnessing elections to clear", count_egress_witnessing);
+
+		let count_nonce_witnessing: u64 = ElectionProperties::<Runtime, SolanaInstance>::iter_keys()
+		.filter(|id| matches!(id.extra(), CompositeElectionIdentifierExtra::C(_)))
+		.count()
+		.try_into()
+		.unwrap();
+		
+		log::info!("🧹 Found {} NonceWitnessing elections to clear", count_nonce_witnessing);
+
+		Ok((count_egress_witnessing, count_nonce_witnessing).encode())
+	}
+
+	fn on_runtime_upgrade() -> Weight {
+		if genesis_hashes::genesis_hash::<Runtime>() == genesis_hashes::BERGHAIN {
+			return Weight::zero();
+		}
+		let next_election_id = if genesis_hashes::genesis_hash::<Runtime>() == genesis_hashes::PERSEVERANCE { 
+			9_397_949u64 
+		} else { 
+			9_346_659u64 
+		};
```

Comment:
nit: It would be a bit safer to have a `match` statement and only ever do anything in the PERSA and SISY branches. The current logic does the right thing, but it will also trigger on localnet, for example.